README.md
gutwahammer
============



Elgg-Hammer ! High Speed Elgg on shared hosting plan or account or VPS or Dedicated server/Hosting is at your fingertips.

Si tu sitio de Elgg desacelerando o retraso o correr mucho más lento de lo normal o normal?

Is your Elgg site slowing down or lagging or running significantly slower than usual or normal?

Online Web site bots, crawlers, spiders, and even spammers can slow down your Elgg site quite significantly or your Elgg site 
loading times might take a little longer than a normal Elgg site. To be honest, the bots, crawlers, spiders, and even spammers 
can make your Elgg site not work like it is suppose to be or like it used to when you first downloaded and installed your fresh 
and new Elgg site. And this is where Elgg Hammer can help make your Elgg site return to its former loading or running speed...

Elgg-Hammer! There are more Elgg hammer function designs to come! If you like it, Recommend it and download it now!

        Automatically ban web site hammers!
        Protect your site against Referer Spam!
        Deny script-kiddie and h4x0r requests!
        Send bad bots and spiders packing!
        Protect your valuable server resources for genuine clients..

 

    ALLOW ONLY SOME GOOD Web robots to search your Elgg site and SHUT others from slowing down your site.
    Allow only some good bots to by pass Elgg Hammer.... altavista, askjeeves, excite, google, infoseek, inktomi, lycos, Microsoft, 
	misc, msn, northernlight, wisenut  ... Anything not in these list will be kicked out or blocked out of any site protected 
	by ELGG HAMMER.

Elgg-Hammer is a php script that runs when your Elgg pages start loading and it searches as requests arrive, Elgg-Hammer checks how 
long it's been since that client's last  request.

If a reasonable amount of time has passed, the page is served as usual. But if not, their "Hammer Count" is increased.

Oh oh ho ho ho! When the elgg hammer trigger levels count reaches, the user if is hammering is immediately suspended, and redirected
 to a page with a message  "must wait X amount of seconds before you will try again!" The more they continue to mess around with your
 website, the longer they will have to wait.

 

If you Like this plugin, consider to click Recommend button for other Elgg members to find it useful for their sites!

Elgg hammer is easy to configure...

 

 HOW TO TEST ELGG HAMMER ON YOUR SITE:

I have received tremendous amount of email on how to test Elgg hammer and for this reason, I have included this short description to help 
those who might have the same question about Elgg hammer.

First, to test Elgg Hammer, you have to install Elgg Hammer "gutwahammer" in your Elgg mod directory, Activate it through Elgg admin plugin 
page, then  load any page of your Elgg website, refresh it repeatedly or hit Key on your keyboard over and over like bots do, quickly. 
 Be careful, if you continue refreshing the page repeatedly I promise you! After a few hits, you will get a banned message on your screen 
 "must wait X amount of seconds before you will try again!" ....  which means you have been banned from hammering your own beautiful elgg site...

 

 

Attributions / Credits

    Elgg Hammer  is an improved and robust version of the Anti-hammer, project at http://corz.org/serv/tools/anti-hammer/

Release Notes:

 

 

Fixed -  Log File: Now Elgg Hammer can  log each banned hit, for site administrator reference purposes.                  

NOTE: It is recommend you watch Elgg Hammer log file very carefully for the first  few minutes/ days after installation and in case of any
 concern, please do e-mail me about it!


changes.txt




	changes:

	0.9.3

		+	You now have the option to perfomr a quick DNS lookup of the
			IP Address of bad clients, and have this added to the logging.

			This was already enabled, you now have the option to *disable* it,
			if required.

		+	Anti-Hammer now send a valid "Retry-After" header, which is set to
			the client's current hammer delay + 1 second.

		+	Added a link to the Anti-Hammer page, should lessen the wtf-factor.



	0.9.2

		+	You can now choose whether to allow your specified clients (aka 
			"exemptions") to  either completely bypass anti-hammer (current
			exemption method)..

				$anti_hammer['allow_bots'] = true;

			Or else specify an integer, representing a hammer_time, in 
			1/100th Second, which will apply to *only* these clients..

				$anti_hammer['allow_bots'] = 50;

			This setting would enable your specified clients to hammer the site 
			at a rate of two hits-per-second, but no faster.

			Effectively, we now have two hammer rates, one for known good 
			clients, and one for everyone else.

	0.9

		+	Good bots & spiders can now be allowed to bypass the hammer. This is
			achieved through the use of standard spider IP lists, as published
			here..

				http://www.iplists.com/

			along with a simple ini file, detailing which user-agent links to
			which IP list. A working ini, and more details, will be included in
			the preference section (above), as well as the release.


	0.8.*

		+	Anti-Hammer now sends a proper 503 (service temporarily unavailable)
			message, rather than a 200 OK message. This will be useful in 
			situations where valid bots are temporarily hammering, and is more
			correct in this scenario. The reource *will* be back, if they cut
			out the crazy hammering!

			If you are running under cgi/*suexec (non-module), the extra 
			required header is automatically sent.

			In use, this causes many bots to back-off immediately. Excellent!


		~	Improved the ban resetting (which needs to work independantly of the
			Garbage collection mechanism). After the ban time, the client's
			cut-off is wiped, and their start time set to *now*, just like a new
			client, however, their hammer_count is set to one hammer below the
			first trigger level. In other words, a single hammer gets them the
			NO Hammer! page; and to the final page quicker than new clients.

			Even if you use rolling triggers, Anti_hammer will still use the 
			first ban level to calculate this number, so set that to whatever 
			you want.

		~	ban_times and ban_levels have been renamed to waiting_times and 
			trigger_levels, to avoid confusion with the ban_time (for the new
			total cut-off). These also make more sense, as they are not bans, 
			simply delays.


	0.7.*

		+	Added rolling ban times. Rather than have set limits which the
			client can cross, this simply increments the ban time with each and 
			every hammer attempt. 1-2-3-4-5-6-7.. etc. cut-off still functions 
			as before for each system (rolling or preset levels).

			This was, in fact, the original system, which I replaced with the
			level presets early on, but it's kinda fun, and the code is simple.

		~	Removed the file-tools.php include statement, and put the functions
			directly into here (slightly renamed). I figure anyone smart enough
			to be including my file-tools, will be smart enough to figure out
			how to put that back, if required. 

		+	More things are configurable, like the page title. Why not!


	0.6.*

		+	Added capability to work with clients who do not accept, or have 
			chosen not to accept (read: disabled) cookies.

			Basically, we write a "fake" session. The fake session uses a 
			serialized array in a flat file, just like regular php sessions, and
			is created before they even get receive their first page. From that
			point on, they are known (by Anti-Hammer) by this ID.

			The name of the client's session ID file is the session ID itself; 
			an MD5 of all the known usable client data concatenated together.

		~	php session usage is still available as an option, if required.

		+	Added Garbage Collection for the fake session files. Both how often 
			this happens (every 'so-many' requests), and how old is considered 
			"stale", are configurable.

		+	Ban time is now configurable (in hours). Remember to ensure that 
			Garbage Collection isn't happening before this time.

		+	Added penultimate message for cut-off. You get one *final* warning!

		~	Cleaned-up the code regarding sessions. We now make a clean break,
			converting whichever type of session data into a local array, and
			then work with that. At the end, we write the pertinent data back to
			whichever type of session is being utilized (built-in or php).


	0.5.*

		+	You can now configure a cut-off point. When the number of violations 
			reaches this number, their pages simply die. This is disabled by 
			default. This point is, of course, configurable. (actually, I got 
			called away in the middle of this, so I'll need to check how far I 
			got!)

	0.4.*

		+	Added user preferences for lots of the settings, voilation levels, 
			times, etc. Added error checking for these, so they should be fairly 
			foolproof (good movie, by the way, "Foolproof", 2003).

	0.3.*

		+	Added configurable protection skipping for certain file types 
			(usually associated files and such). This replaces a nasty hack that 
			lived at the top of the script.

		+	Added skipping for generated images, too (GD images, etc.). This can 
			also be used to skip other tpyes. See the preferences for more 
			details.

		+	Added configurable messages. I'll likely put this out eventually, 
			it's kinda useful.

	0.2.*

		+	Added ignored areas, for chat scripts and such. places where either 
			hammering is allowed, or is dealt with by the local script.



start.php
<?php
/***************************************************************************
	*                            TwizaNex Smart Community Software
	*                            ---------------------------------
	*  Start.php:  gutwacaptcha for Elgg 1.8.15
        *	
	*     begin                : Mon Mar 23 2011
	*     copyright            : (C) 2011 TwizaNex Group
	*     website              : http://www.TwizaNex.com/
	* This file is part of TwizaNex - Smart Community Software
	*
	* @package Twizanex
	* @link http://www.twizanex.com/
	* TwizaNex is free software. This work is licensed under a GNU Public License version 2. 
	* @license http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
	* @author Tom Ondiba <twizanex@yahoo.com>
	* @copyright Twizanex Group 2011
	* TwizaNex is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
	* without even the implied warranty of  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
	* See the GNU Public License version 2 for more details. 
	* For any questions or suggestion write to write to twizanex@yahoo.com
	***************************************************************************/

elgg_register_event_handler('init', 'system', 'gutwahammer_init');

/**
 * Initialize gutwahammer
 */
function gutwahammer_init() {

	global $CONFIG;

	  elgg_register_plugin_hook_handler('route', 'all', 'gutwa_hammer_plugin_hook_handler'); // works but does not triger

}

/**
 * We will grab the existing $page and forward to sef_page_handler, which will call_user_func the original handler
 */


function gutwa_hammer_plugin_hook_handler($hook, $type, $return, $params) {

	$segments = elgg_extract('segments', $return, '');


	$handler = elgg_extract('handler', $return);

// We don't need sef URLs for the following handlers
	$handler_exceptions = array(
		'action',
		'admin',
		'cache',
		'services',
		'export',
		'mt',
		'xml-rpc',
		'rewrite',
		'tag',
		'pg',
		'admin',
		'cron',
		'js',
		'css',
		'ajax',
		'livesearch',
		'activity',
		'setting',
		'friends',
		'friendsof',
		'forgotpassword',
		'resetpassword',
		'login',
		'avatar',
		'profile',
		'collections',	

);

 if (in_array($handler, $handler_exceptions)) {      
          return $return;
         }
       
  // This is our current url
if (sizeof($segments) > 0) {
$segments_url = '/' . implode('/', $segments);
}
$full_url = "{$handler}{$segments_url}";

include_once(dirname(dirname(__FILE__)) . "/gutwahammer/anti-hammer.php");

// return true;
	
}


lib/cleaner.php
<?php

if (realpath($_SERVER['SCRIPT_FILENAME']) == realpath(__FILE__)) { die(
'This script is designed to run if some conditions are ment..<br /><br />
<tt>php_value auto_prepend_file "/real/full/server/path/to/cleaner.php"</tt>'); }


function CleanGarbage() {

// tom addition
 $muda_ngapi = 3600; // How old are the file since they were created
//  $seconds_old = 5;

$dir = dirname(__FILE__).'/requestBlocker';//TM: Added ## Create & set directory writeable!!!! 

if( !$dirtosha = @opendir($dir) )
                        return;

                while( false !== ($chafu = readdir($dirtosha)) ) {
                        if( $chafu != "." && $chafu != ".." ) {
                                $chafu = $dir. "/". $chafu;

                                if( @filemtime($chafu) < (time()-$muda_ngapi) )
                                
                                $twende = @unlink($chafu);


    if($twende=="1"){
        echo "The file was deleted successfully.";
    } else { 
    
        echo "There was an error trying to delete the file.";
    
    
     } 
	
       }
                }
				
				
				}

hammer.php
<?php
//PHP Limit/Block Website requests for Spiders/Bots/Clients etc.

//Here i have written a PHP function which can Block unwanted Requests to reduce your Website-Traffic. God for Spiders, Bots and annoying Clients.

//CLIENT/Bots Blocker

//DEMO: http://szczepan.info/9-webdesign/php/1-php-limit-block-website-requests-for-spiders-bots-clients-etc.html

/* Function which can Block unwanted Requests
 * @return boolean/array status
 */
function requestBlocker()
{
        /*
        Version 1.0 11 Jan 2013
        Author: Szczepan K
        http://www.szczepan.info
        me[@] szczepan [dot] info
        ###Description###
        A PHP function which can Block unwanted Requests to reduce your Website-Traffic.
        God for Spiders, Bots and annoying Clients.

        */

    //    $dir = 'requestBlocker/'; ## Create & set directory writeable!!!! // Original
        $dir = dirname(__FILE__).'/requestBlocker/';//TM: Added ## Create & set directory writeable!!!! 
        
//        echo $dir;

        $rules   = array(
                #You can add multiple Rules in a array like this one here
                #Notice that large "sec definitions" (like 60*60*60) will blow up your client File
                array(
                        //if >5 requests in 5 Seconds then Block client 15 Seconds
                        'requests' => 5, //5 requests
                        'sek' => 5, //5 requests in 5 Seconds
                        'blockTime' => 15 // Block client 15 Seconds
                ),
                array(
                        //if >10 requests in 30 Seconds then Block client 20 Seconds
                        'requests' => 10, //10 requests
                        'sek' => 30, //10 requests in 30 Seconds
                        'blockTime' => 20 // Block client 20 Seconds
                ),
                array(
                        //if >200 requests in 1 Hour then Block client 10 Minutes
                        'requests' => 200, //200 requests
                        'sek' => 60 * 60, //200 requests in 1 Hour
                        'blockTime' => 60 * 10 // Block client 10 Minutes
                )
        );
        
       //  echo $rules;
     //    print_R($rules);
    
    // print_R($client['time']); 
        
        $time    = time();
        $blockIt = array();
        $user    = array();

        #Set Unique Name for each Client-File 
        $user[] = isset($_SERVER['REMOTE_ADDR']) ? $_SERVER['REMOTE_ADDR'] : 'IP_unknown';
        $user[] = isset($_SERVER['HTTP_USER_AGENT']) ? $_SERVER['HTTP_USER_AGENT'] : '';
        $user[] = strtolower(gethostbyaddr($user[0]));

        # Notice that i use files because bots does not accept Sessions
        $botFile = $dir . substr($user[0], 0, 8) . '_' . substr(md5(join('', $user)), 0, 5) . '.txt';


        if (file_exists($botFile)) {
                $file   = file_get_contents($botFile);
                $client = unserialize($file);

        } else {
                $client                = array();
                $client['time'][$time] = 0;
        }

        # Set/Unset Blocktime for blocked Clients
        if (isset($client['block'])) {
                foreach ($client['block'] as $ruleNr => $timestampPast) {
                        $left = $time - $timestampPast;
                        
      //                  print_R($left);
                        
                        if (($left) > $rules[$ruleNr]['blockTime']) {
                                unset($client['block'][$ruleNr]);
                                continue;
                        }
                        $blockIt[] = 'Block active for Rule: ' . $ruleNr . ' - unlock in ' . ($left - $rules[$ruleNr]['blockTime']) . ' Sec.';
                }
                if (!empty($blockIt)) {
                        return $blockIt;
                }
        }

// print_R($ruleNr);

        # log/count each access
        if (!isset($client['time'][$time])) {
                $client['time'][$time] = 1;
        } else {
                $client['time'][$time]++;

        }

        #check the Rules for Client
        $min = array(
                0
        );
        foreach ($rules as $ruleNr => $v) {
                $i            = 0;
                $tr           = false;
                $sum[$ruleNr] = '';
                $requests     = $v['requests'];
                $sek          = $v['sek'];
                foreach ($client['time'] as $timestampPast => $count) {
                        if (($time - $timestampPast) < $sek) {
                                $sum[$ruleNr] += $count;
                                if ($tr == false) {
                                        #register non-use Timestamps for File 
                                        $min[] = $i;
                                        unset($min[0]);
                                        $tr = true;
                                }
                        }
                        $i++;
                }

                if ($sum[$ruleNr] > $requests) {
                        $blockIt[]                = 'Limit : ' . $ruleNr . '=' . $requests . ' requests in ' . $sek . ' seconds!';
                        $client['block'][$ruleNr] = $time;
                }
        }
        $min = min($min) - 1;
        #drop non-use Timestamps in File 
        foreach ($client['time'] as $k => $v) {
                if (!($min <= $i)) {
                        unset($client['time'][$k]);
                }
        }
        $file = file_put_contents($botFile, serialize($client));


        return $blockIt;
        
         echo "go on!"; // for deb urging

}
/*
function CleanGarbage() {

// tom addition
 $muda_ngapi = 3600; // How old are the file since they were created
//  $seconds_old = 5;

$dir = dirname(__FILE__).'/requestBlocker';//TM: Added ## Create & set directory writeable!!!! 

if( !$dirtosha = @opendir($dir) )
                        return;

                while( false !== ($chafu = readdir($dirtosha)) ) {
                        if( $chafu != "." && $chafu != ".." ) {
                                $chafu = $dir. "/". $chafu;

                                if( @filemtime($chafu) < (time()-$muda_ngapi) )
                                
                                $twende = @unlink($chafu);


    if($twende=="1"){
        echo "The file was deleted successfully.";
    } else { 
    
        echo "There was an error trying to delete the file.";
    
    
     } 
	
       }
                }
				
				
}

*/


/*

if ($t = requestBlocker()) {

        header("HTTP/1.1 403 Forbidden");
       exit;
//        echo 'dont pass here!';
//        print_R($t);
} else {
        echo "go on!";
}

*/